uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js

```diff
@@ -0,0 +1,224 @@
+        ]
+      },
+      approve: {
+        label: 'Approve',
+        isDirectAction: true,
+        statuses: [sharedSettings.statusNames.awaitingApproval],
+        dialog: {
+          title: 'Approval',
+          btnClass: 'success',
+          btnLabel: 'Approve',
+          msg: 'Please confirm approval'
+        }
+      },
+      reject: {
+        label: 'Reject',
+        isDirectAction: true,
+        statuses: [sharedSettings.statusNames.awaitingApproval],
+        dialog: {
+          title: 'Rejection',
+          btnClass: 'warning',
+          btnLabel: 'Reject',
+          msg: 'Please confirm rejection'
+        }
+      },
+      cancel: {
+        label: 'Cancel',
+        isDirectAction: true,
+        statuses: [
+          sharedSettings.statusNames.awaitingApproval,
+          sharedSettings.statusNames.moreInformationRequired
+        ],
+        dialog: {
+          title: 'Cancellation',
+          btnClass: 'danger',
+          btnLabel: 'Confirm',
+          msg: 'Please confirm cancellation'
+        }
+      },
+      delete: {
+        label: 'Delete',
+        isDirectAction: true,
+        statuses: [
+          sharedSettings.statusNames.awaitingApproval,
+          sharedSettings.statusNames.moreInformationRequired,
+          sharedSettings.statusNames.approved,
+          sharedSettings.statusNames.rejected,
+          sharedSettings.statusNames.cancelled
+        ],
+        dialog: {
+          title: 'Deletion',
+          btnClass: 'danger',
+          btnLabel: 'Confirm',
+          msg: 'This cannot be undone'
+        }
+      }
+    };
+
+    vm.list = [];
+
+    /**
+     * Performs an action on a given leave request
+     *
+     * @param {LeaveRequestInstance} leaveRequest
+     * @param {string} action
+     */
+    vm.action = function (action) {
+      var dialogParams = actions[action].dialog;
+
+      dialog.open({
+        title: 'Confirm ' + dialogParams.title + '?',
+        copyCancel: 'Cancel',
+        copyConfirm: dialogParams.btnLabel,
+        classConfirm: 'btn-' + dialogParams.btnClass,
+        msg: dialogParams.msg,
+        onConfirm: function () {
+          return vm.leaveRequest[action]()
+            .then(function () {
+              $rootScope.$emit('LeaveRequest::' + (action === 'delete' ? 'deleted' : 'edit'), vm.leaveRequest);
+            });
+        }
+      });
+    };
+
+    init();
+
+    function init () {
+      /* This component expects vm.leaveRequestStatuses and vm.absenceTypes
+       * to be an objects, so if arrays are passed, they are transformed into
+       * object with according indexes
+       */
+      if (Array.isArray(vm.leaveRequestStatuses)) {
+        vm.leaveRequestStatuses = _.indexBy(vm.leaveRequestStatuses, 'value');
+      }
+
+      if (Array.isArray(vm.absenceTypes)) {
+        vm.absenceTypes = _.indexBy(vm.absenceTypes, 'id');
+      }
+
+      var allowedActions = getAllowedActions();
```

Comment:
again, init is doing too many things. It's getting the allowed actions but then it needs to be the one to create the final map. Why don't you assign the value of `vm.list` (or better `vm.allowedActions`) inside the `getAllowedActions` directly?